### PR TITLE
Force menu rebuild when window is closed (#10702)

### LINKED ISSUE #10702: MenuBarItems Doesn't Work Correctly in Multi-Windows on MacCatalyst
### Description

MenuBarItems doesn't show correctly depending on the current active window on MacCatalyst.
Menu bar will always be replaced by the newest window / page even if the window / page is closed.

### Steps to Reproduce

1. Run the project in the given repository.
2. Notice that the menu bar is now showing menus from the main window / page. Example: Main Page Tools.
3. Go to menu bar > Main Page Tools > Open Sub Window.
4. When the new window open, notice that the menu bar is now showing menus from sub window / page instead. Example: Sub Page Tools.
5. Switch back the focus to main window / page without closing the sub window / page.
6. Notice that the menu bar doesn't change back to menus from the main window / page.
7. Now close the sub window / page.
8. Notice that the menu bar still remain unchanged even the sub window / page already closed.

### Link to public reproduction project repository

https://github.com/ooikengsiang/MultiWindowsWithMenu

### Version with bug

7.0 Release Candidate 1

### Last version that worked well

Unknown/Other

### Affected platforms

macOS

### Affected platform versions

MacCatalyst

### Did you find any workaround?

No.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by ooikengsiang**
### Description of Change
Manually resign the key window and force the menu bar rebuild when key window is closed, so that the menu bar in MacOS is shown correctly (if the window is close in the correct order it is created). This changes still doesn't fix menu bar showing incorrectly if user switch window focus in MacOS manually or user close window different than the order it created.

### Issues Fixed
Fixes #10702

### CODE REVIEW

**Review by mandel-macaque — APPROVED** [review]

**Review by PureWeen — CHANGES_REQUESTED** [review]

File: src/Core/src/Platform/iOS/MauiUISceneDelegate.cs [hunk 1]
```diff
@@ -32,6 +32,15 @@ public virtual void WillConnect(UIScene scene, UISceneSession session, UISceneCo
 		public virtual void DidDisconnect(UIScene scene)
 		{
 			MauiUIApplicationDelegate.Current?.Services?.InvokeLifecycleEvents<iOSLifecycle.SceneDidDisconnect>(del => del(scene));
+
+			if (Window is not null && Window.IsKeyWindow)
+			{
+				// manually resign the key window and rebuild the menu
+				Window.ResignKeyWindow();
+				UIMenuSystem
```
[the remark this answers is not in the document]
Comment (ooikengsiang) [reply src/Core/src/Platform/iOS/MauiUISceneDelegate.cs:40]: I agree, but I found a limitation on MacCatalyst when research for this problem. There aren't any API we can pull to get the focused window. It may sound crazy but iPad actually doesn't need this since only one window will be visible at a time which is the latest window created (also key window). The only API that can track focused windows I found is from AppKit (is there a way we can access AppKit?). That is why my fix only can play around with keywindow. Let me know if you have better idea.

[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Core/src/Platform/iOS/MauiUISceneDelegate.cs:40]: I tried this and it seems to work on catalyst
https://stackoverflow.com/a/75440759

I think you can check if the activeappearance on the UIWindows's traitcollection is active.

[the remark this answers is not in the document]
Comment (ooikengsiang) [reply src/Core/src/Platform/iOS/MauiUISceneDelegate.cs:40]: Previously I tried this solution and it doesn't return the focused window but activated scene only. But I was testing many different methods that time and I might be wrong, let me try again and get back to here later.

[the remark this answers is not in the document]
Comment (ooikengsiang) [reply src/Core/src/Platform/iOS/MauiUISceneDelegate.cs]: The active appearance work with some workaround, see my latest changes, now iOS14+ can rebuild menu whenever a window get focused.

File: src/Core/src/Platform/iOS/MauiUIApplicationDelegate.Menu.cs [hunk 2]
```diff
@@ -24,9 +26,43 @@ public override void BuildMenu(IUIMenuBuilder builder)
 
 			MenuBuilder = builder;
 
-			var window = Window ?? this.GetWindow() ??
-				UIApplication.SharedApplication.GetWindow()?.Handler?.PlatformView as UIWindow;
-
+			UIWindow? window = null;
+			if (OperatingSystem.IsIOSVersionAtLeast(14))
```
Comment (PureWeen) [inline src/Core/src/Platform/iOS/MauiUIApplicationDelegate.Menu.cs:30]:
```suggestion
			if (OperatingSystem.IsMacCatalystVersionAtLeast(14))
```

File: src/Core/src/Platform/iOS/MauiUIApplicationDelegate.Menu.cs [hunk 3]
```diff
@@ -1,4 +1,6 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (PureWeen) [inline src/Core/src/Platform/iOS/MauiUIApplicationDelegate.Menu.cs:3]: We try to stay away from using LINQ with any core code because it can occasionally lead to some performance issues. Can you rework the code to not use linq?

File: src/Core/src/Platform/iOS/MauiUISceneDelegate.cs [hunk 4]
```diff
@@ -32,6 +32,20 @@ public virtual void WillConnect(UIScene scene, UISceneSession session, UISceneCo
 		public virtual void DidDisconnect(UIScene scene)
 		{
 			MauiUIApplicationDelegate.Current?.Services?.InvokeLifecycleEvents<iOSLifecycle.SceneDidDisconnect>(del => del(scene));
+
+			// for iOS 13 only where active apperance is not supported yet
+			// for iOS 14+, see DidUpdateCoordinateSpace
+			if (OperatingSystem.IsIOSVersionAtLeast(13) && !OperatingSystem.IsIOSVersionAtLeast(14))
```
Comment (PureWeen) [inline src/Core/src/Platform/iOS/MauiUISceneDelegate.cs:38]:
```suggestion
			if (!OperatingSystem.IsMacCatalystVersionAtLeast(14))
```

**Review by PureWeen — CHANGES_REQUESTED** [review]
Few comments. This PR is amazing thank you! Apologies on the slow churn. I tested this out with a few different permutations of windows and menus and it worked perfectly!!!

**Comment by ooikengsiang** [comment]
> Few comments. This PR is amazing thank you! Apologies on the slow churn. I tested this out with a few different permutations of windows and menus and it worked perfectly!!!

Thanks for the review. I have updated the code base on your review.

**Review by PureWeen — APPROVED** [review]
